scripts/docker-run.sh

```diff
@@ -38,6 +38,19 @@ else
 	DOCKER_RUN="docker run"
 fi
 
+#
+# Ensure that UPSTREAM_PRODUCT_BRANCH is set.
+#
+if [[ -z "$UPSTREAM_PRODUCT_BRANCH" ]]; then
```

Comment:
What is expected when an incorrect value is specified? If “master” is set, but we’re building a non-master branch, will the build fail, or silently do the wrong thing?

Also, please update the build instructions in the README to document this new requirement.

Reply:
If the wrong values is specified then the job will silently do the wrong thing. It will attempt the build, but pull in the incorrect packages for that version. I think the build will succeed, but with those wrong package versions.

Reply:
I want to point out where this is likely to do the wrong thing...

In Jenkins, we configure the jobs that is used by "bors" like this:
```
createApplianceBuildJob('post-push', 'staging', 'master', false, false, false, false) 
```
i.e. we assume that anytime we're building the "staging" branch, the "upstream branch" is "master".

This assuming is not true if we're attempting to "bors r+" a pull request that is not based on master. For example, a backport PR will be based on the "6.0/release" branch, and when "bors r+" is used, the "staging" branch will become the 6.0/release branch merged with the pull request.

Based on this logic in this change, coupled with how our Jenkins logic incorrectly assumes "staging" always maps to "master", we'd end up doing a build of the 6.0/release branch, but wind up using the "master" branch's package repository.

Arguably, this is a bug in the Jenkins logic, and this change would simply be a victim here; but regardless, I think it's worth pointing out this issue, and being aware of it before landing this change.

Reply:
I've opened https://jira.delphix.com/browse/TOOL-8444 to track this issue.

Reply:
I think with this concern and the concern that if a developer specifies an incorrect UPSTREAM_PRODUCT_BRANCH the build will silently be done with an incorrect mirror dataset I am going to reintroduce the hard coded BRANCH parameter. I think I am going to add this at a file in the root of the repo called something like `branch.config` to make it easy for the release lead to update this parameter, and make it easier to possibly automate this change later.

With this concern with bors I think the UPSTREAM_PRODUCT_BRANCH parameter should just be pulled out, and appliance build should always just pull in the BRANCH from the new branch config file. Can you think of a situation where a developer would want to override this? They could do that by changing the value in the config file, but I am trying to decide if this should be a documented workflow.

Reply:
I think given the issues this seems reasonable. 

I'd also like to have feedback from other release-team folks like @shartse and @basil .

Reply:
I'd be fine using some sort of "metadata file", where we specify this sort of information.

We do something similar to this with the `.hooksconfig` file in some of our other repositories. There's also some precedence for this kind of approach in the upstream ZFS on Linux code base; see [here](https://github.com/delphix/zfs/blob/master/META).

So, as a general design, I think it could work. If we go down this route, I think it'd be best to try and design a solution that could generally be used in all of our various repositories, so we have a common solution across them all (potentially, at least; but only if the specific repositories requires the complexity).

This is something that'd need to be updated when we create new release branches; but if we're consistent, we could probably write some automation to create the branch and update the file at the same time.

Reply:
> They could do that by changing the value in the config file, but I am trying to decide if this should be a documented workflow.

IMO, I wouldn't document it, nor officially support changing the value.

Reply:
👍